monitor/serie.cc

```diff
@@ -121,51 +121,98 @@ namespace rtm
+                                 (limits.X.Max <= cached_max_);
 
-            // display 3 sections (before, here and after if possible)
-            if (it != sections_.begin())
+            if (!can_use_cache)
             {
-                it--;
-            }
+                // Cache miss - need to find the sections
+                cached_sections_.clear();
 
-            for (int i = 0; i < 3; ++i)
-            {
-                Parser::Point const* data = it->points.data();
-                ImPlot::PlotLine(name_.c_str(), &data->x, &data->y, static_cast<int>(it->points.size()),
-                    0, 0, sizeof(Parser::Point));
+                // Find first overlapping section
+                auto it_first = std::find_if(sections_.begin(), sections_.end(), [&](Section const &s)
+                                             { return (s.max.count() >= limits.X.Min) && (s.min.count() <= limits.X.Max); });
+
+                if (it_first == sections_.end())
+                {
+                    // No sections overlap - fallback to downsampled view
+                    cache_valid_ = false;
+                    ImPlot::PlotLine(name_.c_str(), &serie_.at(0).x, &serie_.at(0).y,
+                                     static_cast<int>(serie_.size()), 0, 0, sizeof(Parser::Point));
+                    return is_downsampled_;
+                }
+
+                // Add one section before if possible (for smooth scrolling)
+                if (it_first != sections_.begin())
+                {
+                    --it_first;
+                }
+
+                // Collect all overlapping sections + neighbors
+                auto it = it_first;
+                while (it != sections_.end() &&
+                       (cached_sections_.empty() ||
+                        it->min.count() <= limits.X.Max + SECTION_SIZE.count()))
+                {
+                    cached_sections_.push_back(&(*it));
+                    ++it;
+                }
+
+                // Update cache bounds with some margin for small movements
+                double view_size = limits.X.Size();
+                double margin = view_size * CACHE_MARGIN_RATIO;
+
+                cached_min_ = cached_sections_.front()->min.count() - margin;
+                cached_max_ = cached_sections_.back()->max.count() + margin;
+                cache_valid_ = true;
 
-                it++;
-                if (it == sections_.end())
+                // To Remove once PR validated
+                printf("Cache MISS - Found %zu sections covering [%.2f, %.2f] with cache [%.2f, %.2f]\n",
+                       cached_sections_.size(),
+                       cached_sections_.front()->min.count(),
+                       cached_sections_.back()->max.count(),
+                       cached_min_, cached_max_);
+            }
+            else
+            {
+                // To Remove once PR validated
+                static int hit_count = 0;
```

Comment:
not sure if your stats are good since you don't use a member (even for debug)

Reply:
this was dirty debugging i wil remove it